requests/utils.py

```diff
@@ -87,7 +83,22 @@ def super_len(o):
             # is actually a special file descriptor like stdin. In this
             # instance, we don't know what the length is, so set it to zero and
             # let requests chunk it instead.
-            current_position = total_length
+            if total_length is not None:
+                current_position = total_length
+
+        if hasattr(o, 'seek') and total_length is None:
+            # StringIO and BytesIO have seek but no useable fileno
+
+            # seek to end of file
+            o.seek(0, 2)
+            total_length = o.tell()
+
+            # seek back to current position to support
+            # partially read file-like objects
+            o.seek(current_position or 0)
```

Comment:
So we set current_position on [line 80](https://github.com/nateprewitt/requests/blob/03535ce5baa0931cdce01b849ca37795229f3bc5/requests/utils.py#L80) to o.tell() which in most cases should be a valid number. If someone defines a custom tell() method that returns None under circumstance x, this will blow up with a `TypeError`.

I think this is a pretty rare edge case but I understand why @jseabold put it in place. Why leave a point of failure to avoid a quick comparison?

I'll defer to @Lukasa if he has an opinion, otherwise I think it's worth keeping.